src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,439 @@
+      ... ≥ ∥P(Q x)∥ + ∥(Q x - P (Q x)) + (x - Q x)∥ :
+        by apply (add_le_add_iff_left (∥P(Q x)∥)).mpr (norm_add_le (Q x - P (Q x)) (x - Q x))
+      ... = ∥P(Q x)∥ + ∥x - P (Q x)∥ : by rw sub_add_sub_cancel'
+      ... = ∥(P * Q) x∥ + ∥(1 - P * Q) x∥ : rfl }, }
+end
+
+lemma join {P Q: X →L[𝕜] X} (h₁ : is_Lprojection P) (h₂ : is_Lprojection Q) :
+  is_Lprojection (P + Q - P * Q) :=
+begin
+  have e1:  1 - (1 - P) * (1 - Q) = P + Q - P * Q :=
+  calc 1 - (1 - P) * (1 - Q) = 1 -(1 - Q - P * (1 - Q)) : by rw [sub_mul, one_mul]
+  ... = Q + P * (1 - Q) : by rw [sub_sub, sub_sub_self]
+  ... = P + Q - P * Q : by rw [mul_sub, mul_one, add_sub, add_comm],
+  rw ← e1,
+  rw ← is_Lprojection.Lcomplement_iff,
+  apply is_Lprojection.mul,
+  apply is_Lprojection.Lcomplement h₁,
+  apply is_Lprojection.Lcomplement h₂,
+end
+
+instance : has_compl { f : X →L[𝕜] X // is_Lprojection f } :=
+⟨λ P, ⟨1-P, P.prop.Lcomplement⟩⟩
+
+@[simp] lemma coe_compl (P : subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :
+  ↑(Pᶜ) = (1:X →L[𝕜] X) - ↑P := rfl
+
+instance : has_inf (subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :=
+⟨λ P Q, ⟨P * Q, P.prop.mul Q.prop⟩ ⟩
+
+@[simp] lemma coe_inf (P Q : subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :
+  ↑(P ⊓ Q) = ((↑P : (X →L[𝕜] X)) * ↑Q) := rfl
+
+instance : has_sup (subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :=
+⟨λ P Q, ⟨P + Q - P * Q, P.prop.join Q.prop⟩ ⟩
+
+@[simp] lemma coe_sup (P Q : subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :
+  ↑(P ⊔ Q) = ((↑P:X →L[𝕜] X) + ↑Q - ↑P * ↑Q) := rfl
+
+instance : has_sdiff (subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :=
+⟨λ P Q, ⟨P * (1-Q), by exact is_Lprojection.mul P.prop (is_Lprojection.Lcomplement Q.prop) ⟩⟩
+
+@[simp] lemma coe_sdiff (P Q : subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :
+  ↑(P \ Q) = (↑P:X →L[𝕜] X) * (1-↑Q) := rfl
+
+instance : partial_order (subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :=
+{ le := λ P Q, (↑P:X →L[𝕜] X) = ↑(P ⊓ Q),
+  le_refl := λ P, begin
+    simp only [coe_inf],
+    rw [← sq, projection_def],
+    exact P.prop.left,
+  end,
+  le_trans := λ P Q R, begin
+    intros h₁ h₂,
+    simp only [coe_inf],
+    have e₁: ↑P = ↑P * ↑Q := h₁,
+    have e₂: ↑Q = ↑Q * ↑R := h₂,
+    have e₃: (↑P:X →L[𝕜] X) = ↑P * ↑R := begin
+      nth_rewrite_rhs 0 e₁,
+      rw [mul_assoc, ← e₂, ← e₁],
+    end,
+    apply e₃,
+  end,
+  le_antisymm := λ P Q, begin
+    intros h₁ h₂,
+    have e₁: ↑P = ↑P * ↑Q := h₁,
+    have e₂: ↑Q = ↑Q * ↑P := h₂,
+    have e₃: (↑P:X →L[𝕜] X) = ↑Q := by rw [e₁, commute.eq (Lproj_commute P.prop Q.prop), ← e₂],
+    apply subtype.eq e₃,
+  end, }
+
+instance : has_zero (subtype (is_Lprojection  : (X →L[𝕜] X) → Prop))  :=
+⟨⟨0, begin
+  unfold is_Lprojection,
+  split,
+  { unfold is_projection,
+    rw [sq, zero_mul], },
+  { intro, simp only [continuous_linear_map.zero_apply, norm_zero, sub_zero,
+      continuous_linear_map.one_apply, zero_add], },
+end⟩⟩
```

Comment:
```suggestion
instance : has_zero {P : X →L[𝕜] X // is_Lprojection P}  :=
⟨⟨0, ⟨by rw [is_projection, sq, zero_mul],
     λ x, by simp only [continuous_linear_map.zero_apply, norm_zero, sub_zero,
                        continuous_linear_map.one_apply, zero_add]⟩⟩⟩
```
Note in particular the notation for subtypes, this is usually the preferred way to write this.

Reply:
Does this mean that one should have e.g.
```
@[simp] lemma coe_compl (P : {P : X →L[𝕜] X // is_Lprojection P}) :
  ↑(Pᶜ) = (1:X →L[𝕜] X) - ↑P := rfl
```
Is there any way of making the inner `P` anonymous?